packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,51 @@
+export interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+export const MAX_MAP_ENTIRES = 50;
+
+export default function Memoize(resolver?: Function) {
+  return function(...args: any[]) {
+    return Memoized(args[0], resolver);
+  };
+}
+
+function Memoized(func: Function, resolver?: Function) {
```

Comment:
These types are much too weak, the returned function should be the same type as the input one, and the resolver function type can be inferred from the arguments of the memoized function